### PR TITLE
codec: canal-json support batch decoder

### PR DESCRIPTION
**Pull request by 3AceShowHand**
### What problem does this PR solve? 

This is a partwork of making Kafka Sink GA, support batch decoder interface for Canal-JSON format.

### What is changed and how it works?

* implement the `EventBatchDecoder` for Canal-JSON protocol.

### Check List 

Tests 

 - Unit test

Code changes

Side effects

Related changes

### Release note 

```release-note
Please add a release note.
If you don't think this PR needs a release note then fill it with `None`.
```

### CODE REVIEW

File: cdc/sink/producer/kafka/kafka.go [hunk 1]
```diff
@@ -457,18 +457,20 @@ func topicPreProcess(topic string, protocol codec.Protocol, config *Config, sara
...
 	// when try to create the topic, we don't know how to set the `max.message.bytes` for the topic.
 	// Kafka would create the topic with broker's `message.max.bytes`,
-	// we have to make sure it's not greater than `max-message-bytes`
-	brokerMessageMaxBytes, err := getBrokerMessageMaxBytes(admin)
-	if err != nil {
-		log.Warn("TiCDC cannot find `message.max.bytes` from broker's configuration")
-		return errors.Trace(err)
-	}
+	// we have to make sure it's not greater than `max-message-bytes` for the default open protocol & craft protocol.
+	if protocol == codec.ProtocolDefault || protocol == codec.ProtocolCraft {
```
Comment (amyangfei) [inline cdc/sink/producer/kafka/kafka.go:461]: Is this change related to https://github.com/pingcap/ticdc/pull/3192?

Comment (3AceShowHand) [reply cdc/sink/producer/kafka/kafka.go:461]: Yes. I was thinking about should we check `max-message-bytes` < `message.max.bytes` for protocol which not batching messages, such as Canal-JSON, and I think it's unnecessary, because most of the time, the Canal-JSON message should be successfully sent, and once meet size too large problem, it's unrelated to `max-message-bytes`, so only check for open protocol and craft.

File: cdc/sink/codec/canal_flat.go [hunk 2]
```diff
@@ -362,3 +441,146 @@ func (c *CanalFlatEventBatchEncoder) SetParams(params map[string]string) error {
...
 }
+
+// CanalFlatEventBatchDecoder decodes the byte into the original message.
+type CanalFlatEventBatchDecoder struct {
+	data                []byte
+	msg                 *MQMessage
+	enableTiDBExtension bool
+}
+
+func NewCanalFlatEventBatchDecoder(data []byte, enableTiDBExtension bool) (EventBatchDecoder, error) {
```
Comment (lonng) [inline cdc/sink/codec/canal_flat.go:452]: Seems the function `NewCanalFlatEventBatchDecoder` never returns an error, but why do you define your signature like this?

File: cdc/sink/codec/canal_flat_test.go [hunk 3]
```diff
@@ -90,6 +98,85 @@ func (s *canalFlatSuite) TestNewCanalFlatMessageFromDML(c *check.C) {
...
+	encodedBytes, err := charmap.ISO8859_1.NewDecoder().Bytes([]byte("测试blob"))
+	c.Assert(err, check.IsNil)
+	expected := map[string]interface{}{
+		"id":         "1",
+		"name":       "Bob",
+		"tiny":       "255",
+		"comment":    "测试",
+		"blob":       string(encodedBytes),
+		"journalist": "Sharon run so fast",
+		"elder":      "图样图森破",
```
Comment (lonng) [inline cdc/sink/codec/canal_flat_test.go:113]: Prefer to use more common non-English characters.

File: cdc/sink/codec/canal_flat.go [hunk 4]
```diff
@@ -362,3 +441,146 @@ func (c *CanalFlatEventBatchEncoder) SetParams(params map[string]string) error {
...
+
+	return result
+}
+
+func canalFlatJSONColumnMap2SinkColumns(cols map[string]interface{}, mysqlType map[string]string) []*model.Column {
+	result := make([]*model.Column, 0, len(cols))
+	for name, value := range cols {
+		typeStr, ok := mysqlType[name]
+		if !ok {
+			log.Panic("mysql type does not found", zap.String("column name", name), zap.Any("mysqlType", mysqlType))
```
Comment (amyangfei) [inline cdc/sink/codec/canal_flat.go:551]: Can this return error, instead of panic.
Supposing TiCDC really meets panic here, the replication will pause, TiCDC nodes will panic one by one (given that the table will be rescheduled.

File: cdc/sink/codec/canal_flat.go [hunk 5]
```diff
@@ -39,6 +42,36 @@ type CanalFlatEventBatchEncoder struct {
 
 const tidbWaterMarkType = "TIDB_WATERMARK"
 
+var str2MySQLType = map[string]byte{
```
[the remark this answers is not in the document]
Comment (3AceShowHand) [reply cdc/sink/codec/canal_flat.go:45]: At the moment, it's a reverse map of the `https://github.com/pingcap/tidb/blob/master/parser/types/etc.go#L43`.

But it would be hard to track TiDB type changes. If we would update the TiDB dependency in the go mod file, we have to manually check for this.

[the remark this answers is not in the document]
Comment (overvenus) [reply cdc/sink/codec/canal_flat.go:45]: How about exporting `type2Str` in parser and reversing map at runtime?

**Review by overvenus — COMMENTED** [review]
Left one comment https://github.com/pingcap/ticdc/pull/3203/files#r750851191, rest LGTM

File: cdc/sink/codec/canal_flat.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (3AceShowHand) [reply cdc/sink/codec/canal_flat.go:45]: If we can find a way to only reversing the map once, I think exporting `type2Str` is a good choice, otherwise, I would prefer move `str2MySQLType` to the parser directly.

**Review by amyangfei — APPROVED** [review]

File: cdc/sink/producer/kafka/kafka.go [hunk 6]
```diff
@@ -457,18 +457,20 @@ func topicPreProcess(topic string, protocol codec.Protocol, config *Config, sara
 
 	// when try to create the topic, we don't know how to set the `max.message.bytes` for the topic.
 	// Kafka would create the topic with broker's `message.max.bytes`,
-	// we have to make sure it's not greater than `max-message-bytes`
-	brokerMessageMaxBytes, err := getBrokerMessageMaxBytes(admin)
-	if err != nil {
-		log.Warn("TiCDC cannot find `message.max.bytes` from broker's configuration")
-		return errors.Trace(err)
-	}
+	// we have to make sure it's not greater than `max-message-bytes` for the default open protocol & craft protocol.
```
Comment (liuzix) [inline cdc/sink/producer/kafka/kafka.go:460]: Can you move these to a separate PR? Rest LGTM

**Review by overvenus — APPROVED** [review]
LGTM
